smolyay/surrogate.py

```diff
@@ -1,546 +1,742 @@
+
+        # create lookup table and solve for all the basis functions
+        lookup_table = [
+            self.basis_sets[dim](X[:, dim], self.domain[dim])
+            for dim in range(self.num_dimensions)
+        ]
+
+        # create basis matrix
+        if any(
+            any(bf._is_complex for bf in basis_set) for basis_set in self.basis_sets
+        ):
+            basis_matrix = numpy.zeros(
+                (len(X), len(self._index_combinations)), dtype="complex_"
+            )
+        else:
+            basis_matrix = numpy.zeros((len(X), len(self._index_combinations)))
+
+        # use lookup table to solve for each term
+        for term, ic in enumerate(self._index_combinations):
+            basis_matrix[:, term] = numpy.prod(
+                [lookup_table[dim][ic[dim], :] for dim in range(len(ic))], axis=0
+            )
+
+        # solve for coefficients
+        if self.regularization is None or numpy.any(numpy.iscomplex(basis_matrix)):
+            if basis_matrix.shape[0] == basis_matrix.shape[1]:
+                try:
+                    self._coefficients = numpy.linalg.solve(basis_matrix, y)
+                except numpy.linalg.LinAlgError:
+                    self._coefficients = numpy.linalg.lstsq(
+                        basis_matrix, y, rcond=None
+                    )[0]
             else:
-                value = basis(points)
-            basis_matrix[:,j]= value
-
-        if linear_solver == 'lu':
-            self._coefficients = numpy.linalg.solve(
-                basis_matrix,
-                self._data)
-        elif linear_solver == 'inv':
-            self._coefficients = numpy.dot(numpy.linalg.inv(
-                basis_matrix),
-                self._data)
-        elif linear_solver == 'lstsq':
-            self._coefficients = numpy.linalg.lstsq(basis_matrix,
-                                                    self._data,
-                                                    rcond=None)[0]
+                self._coefficients = numpy.linalg.lstsq(basis_matrix, y, rcond=None)[0]
+        elif isinstance(self.regularization, L2Regularization):
+            regressor = sklearn.linear_model.Ridge(**self.regularization.__dict__)
+            self._coefficients = numpy.squeeze(regressor.fit(basis_matrix, y).coef_)
+        elif isinstance(self.regularization, L1Regularization):
+            regressor = sklearn.linear_model.Lasso(**self.regularization.__dict__)
+            self._coefficients = numpy.squeeze(regressor.fit(basis_matrix, y).coef_)
         else:
-            raise ValueError('Solver not recognized')
+            self._coefficients = numpy.linalg.lstsq(basis_matrix, y, rcond=None)[0]
+        self._valid_cache = True
+        self._fit_gradient_flag = False
+        return self
 
-    @staticmethod
-    def _mapdomain(x, old, new):
-        """Transform the points into new domain.
+    def fit_gradient(self, X, y, X0=None, y0=None):
```

Comment:
Currently, `X0` can be set without `y0` and nothing happens. Should we specify a default `y0` of zero, as a default y-value for shifting?
```suggestion
    def fit_gradient(self, X, y, X0=None, y0=0):
```

Reply:
No, because that means the method assumes by default that we know that the lower domain point is 0, whereas before it assumed we knew nothing. Shifting should only happen if the user specifically tells the method to do it.

I don't see an issue with it doing nothing if X0 is set without y0, but if that is a problem, I can add a check for if X0 is not None and y0 is None and have it assume y0 = 0 in that specific case.

Reply:
Oh OK, yes, you are right. I think no change is needed.